### PR TITLE
Make it possible to remove a closed IO::Socket handle from IO::Select

### LINKED ISSUE #17447: IO::Select can't remove closed socket [rt.cpan.org #77740]
Migrated from [rt.cpan.org#77740](https://rt.cpan.org/Ticket/Display.html?id=77740) (status was 'open')

Requestors:
* bitcard@stormcloud9.net

Attachments:
* [rt77740.patch](https://rt.cpan.org/Ticket/Attachment/1328962/704950/rt77740.patch)

From bitcard@stormcloud9.net on 2012-06-09 23:13:45:
If an IO::Socket handle is closed before removing it from the IO::Select  object, IO::Select is unable to remove it.

```perl
use IO::Socket;
use IO::Select;
my $sock = IO::Socket::INET->new('PeerAddr' => 'google.com', 'PeerPort' 
=> 80, 'Proto' => 'tcp');
my $ios = IO::Select->new();
$ios->add($sock);
close($sock);
print("Removing handle $sock\n");
$ios->remove($sock); # <-- this does not work
print("Remaining handles: " . join(" ", $ios->handles()) . "\n");
```

From leonerd-cpan@leonerd.org.uk on 2014-02-20 18:28:43:

Attached patch fixes this.

-- 

Paul Evans

### PR DESCRIPTION
**Pull request by toddr**
Fixes #17447

### CODE REVIEW

**Comment by toddr** [comment]
travis failed on the new test with:

```
syswrite() isn't allowed on :utf8 handles at ../../lib/IO/Handle.pm line 461.
# Looks like your test exited with 255 before it could output anything.
../dist/IO/t/gh17447.t ............................................. 
Dubious, test returned 255 (wstat 65280, 0xff00)
Failed 2/2 subtests 
```

**Comment by tonycoz** [comment]
> 
> 
> travis failed on the new test with:
> 
> ```
> syswrite() isn't allowed on :utf8 handles at ../../lib/IO/Handle.pm line 461.
> # Looks like your test exited with 255 before it could output anything.
> ../dist/IO/t/gh17447.t ............................................. 
> Dubious, test returned 255 (wstat 65280, 0xff00)
> Failed 2/2 subtests 
> ```

The failing travis cases set LANG to a UTF-8 locale and PERL_UNICODE to '', combined they make files/handles open with :utf8 by default.

binmode() on the handles (in the test code) will fix it.

Otherwise it seems sane to me.

File: dist/IO/lib/IO/Select.pm [hunk 1]
```diff
@@ -57,7 +57,21 @@ sub _fileno
  my($self, $f) = @_;
  return unless defined $f;
  $f = $f->[0] if ref($f) eq 'ARRAY';
- ($f =~ /^\d+$/) ? $f : fileno($f);
+ if($f =~ /^\d+$/) { # plain file number
```
Comment (toddr) [inline dist/IO/lib/IO/Select.pm]: Shouldn't this be `/^[0-9]$/`

Comment (shlomif) [reply dist/IO/lib/IO/Select.pm]: Hi! You likely meant `/^[0-9]+$/` (with the `+`) and you are right: https://perl-begin.org/tutorials/bad-elements/#using-character-classes-with-unicode-text .

Comment (toddr) [reply dist/IO/lib/IO/Select.pm]: Yep. Fixed and re-smoking.

**Pull request merged**